## LINEAR REGRESSION - HR Analytics
---

### Assumptions made during Linear Regression Analysis-

####     1)Linearity                 : The relationship between x & the mean of y is Linear.
####     2)Homoscedasticity : Constant Variance.
####     3)Independence        : Observations must independent to each other
####    4)Normality                 : For any fixed value of x & y is Normally Distributed 

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

In [4]:
data= pd.read_csv(r'C:\Users\user\Desktop\skill Enable Tutorial\Python\CTCdata.csv')

data.head()

CTCoffered  LastCTC  Interview rating  Skill Set Index  \
0          19       18                 4                3   
1          17       16                 4                3   
2          17       16                 4                3   
3           9        8                 3                1   
4          10        9                 5                4   

   Highest qualification  Total years of work exp  
0                      3                      8.5  
1                      3                      7.7  
2                      3                      7.9  
3                      2                      2.7  
4                      4                      9.7

In [6]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 191 entries, 0 to 190
Data columns (total 6 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   CTCoffered               191 non-null    int64  
 1   LastCTC                  191 non-null    int64  
 2   Interview rating         191 non-null    int64  
 3   Skill Set Index          191 non-null    int64  
 4   Highest qualification    191 non-null    int64  
 5   Total years of work exp  191 non-null    float64
dtypes: float64(1), int64(5)
memory usage: 9.1 KB


In [7]:
data.isnull().sum()

CTCoffered                 0
LastCTC                    0
Interview rating           0
Skill Set Index            0
Highest qualification      0
Total years of work exp    0
dtype: int64

In [9]:
df= data[['CTCoffered','LastCTC']]
df.head()

CTCoffered  LastCTC
0          19       18
1          17       16
2          17       16
3           9        8
4          10        9

In [10]:
# Identifying independent and dependent variable
x = df[['LastCTC']]    # Independent varibale
y = df[['CTCoffered']] # dependent variable

In [12]:
x.head()

LastCTC
0       18
1       16
2       16
3        8
4        9

In [13]:
y.head()

CTCoffered
0          19
1          17
2          17
3           9
4          10

## Splitting the data for traning and testing

In [14]:
from sklearn.model_selection import train_test_split

In [16]:
x_train,x_test,y_train,y_test = train_test_split(x,y,train_size =0.8,random_state=42)
# Make sure train size should be more(>80%) for better results

In [17]:
len(x_train),len(x_test),len(y_train),len(y_test)

(152, 39, 152, 39)

## Model Building 

In [20]:
from sklearn.linear_model import LinearRegression
lr=LinearRegression()
model=lr.fit(x_train,y_train)

In [21]:
model.coef_   ### Slope(m)

array([[0.9401281]])

In [22]:
model.intercept_  ### Intercept(c)

array([1.80548038])

In [23]:
model.score(x_train,y_train) ### R2 Which tells the how much % model has been able to learn.

0.9709568208423041

In [24]:
y_test.head()


CTCoffered
152          18
75           21
158           8
66           17
60           17

### prediction by using test data

In [25]:
y_test['Predicted CTC'] = model.predict(x_test)

In [26]:
y_test.head()

CTCoffered  Predicted CTC
152          18      17.787658
75           21      20.608042
158           8       7.446249
66           17      16.847530
60           17      16.847530

### Evaluating The Model by Finding out the RMSE

In [27]:
y_test['Error']=y_test['CTCoffered']- y_test['Predicted CTC']

In [28]:
y_test.head()

CTCoffered  Predicted CTC     Error
152          18      17.787658  0.212342
75           21      20.608042  0.391958
158           8       7.446249  0.553751
66           17      16.847530  0.152470
60           17      16.847530  0.152470

In [29]:
y_test['Squared_Error']= y_test['Error']**2

In [30]:
y_test.head()

CTCoffered  Predicted CTC     Error  Squared_Error
152          18      17.787658  0.212342       0.045089
75           21      20.608042  0.391958       0.153631
158           8       7.446249  0.553751       0.306640
66           17      16.847530  0.152470       0.023247
60           17      16.847530  0.152470       0.023247

In [31]:
mse=y_test['Squared_Error'].mean()

In [32]:
mse    ### Mean Squared Error

0.5104422481086017

In [33]:
from math import sqrt
rmse =sqrt(mse)

In [34]:
rmse  ### Root Mean Squared Error

0.7144524113673364

In [35]:
predicted_sal=model.predict([[14]])

C:\Users\user\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


In [36]:
predicted_sal   

array([[14.96727382]])

In [37]:
upperband = predicted_sal+rmse
lowerband = predicted_sal-rmse

In [38]:
upperband


array([[15.68172623]])

In [39]:
lowerband

array([[14.25282141]])

### For 14 Lac CTC The range of CTC that can offered to the candidate is 14.25 - 15.681 Lac